### PR TITLE
feat: Add top-k evaluator, and new models (LSA, LDA, TFIDF)

### PR DESCRIPTION
**Pull request by xhlulu**
## Description

* Adding a new evaluation script that would print the top-k accuracy of a model instead of the top-1 accuracy. 
* Added new models based on topic modelling (LSA,LDA) and bag of words. Each new model has a method `topk(self, question, k=10)` that will output the top k answers based on the question.
* Created a new version of ES (called es_enhanced) that has the `topk` method.

## Files Changed

* `covidfaq/evaluating/model/__init__.py `: Added an utility function for isolating punctuations.

## New Files

* `covidfaq/evaluating/model/es_enhanced_reranker.py`: The same model as the original Elastic Search, but with a new method.
* `covidfaq/evaluating/model/lsa.py`: Latent semantic analysis. The answer relevance score is based on cosine similarity between the latent semantics of the questions and all the answers. It is very fast, and the top-k results are better than Elastic Search.
* `covidfaq/evaluating/model/lda.py`: Latent Dirichlet allocation with cosine similarity. It's slow, and the results do not seem very good. 
* `covidfaq/evaluating/model/tfidf.py`: TFIDF with cosine similarity. Exactly same result as LSA, but it's likely more expensive to compute since it is very high dimensional.

## Results

Top-K Results for Elastic Search:
```
INFO:__main__:top-1 evaluation: correct 133 over 313 / accuracy: 42.49%
INFO:__main__:top-5 evaluation: correct 188 over 313 / accuracy: 60.06%
INFO:__main__:top-10 evaluation: correct 203 over 313 / accuracy: 64.86%
INFO:__main__:top-20 evaluation: correct 212 over 313 / accuracy: 67.73%
INFO:__main__:top-50 evaluation: correct 236 over 313 / accuracy: 75.40%
```

Top-K results for LSA:
```
INFO:__main__:top-1 evaluation: correct 133 over 313 / accuracy: 42.49%
INFO:__main__:top-5 evaluation: correct 211 over 313 / accuracy: 67.41%
INFO:__main__:top-10 evaluation: correct 238 over 313 / accuracy: 76.04%
INFO:__main__:top-20 evaluation: correct 261 over 313 / accuracy: 83.39%
INFO:__main__:top-50 evaluation: correct 286 over 313 / accuracy: 91.37%
```

Results for TFIDF are the same as LSA. LDA show very poor results and was omitted.

## Checklist

- [ ] PR title follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) guidelines 
- [ ] All relevant PR sections are populated, irrelevant ones are removed

### CODE REVIEW

File: covidfaq/evaluating/model/es_enhanced_reranker.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import logging
```
Comment (mirkobronzi) [inline covidfaq/evaluating/model/es_enhanced_reranker.py:1]: what is the difference between this ES and the other?
I think it's better if we make this one subclass of the other and we only change the code related to the differences..

File: covidfaq/evaluating/model/lda.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import numpy as np
+from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
+from sklearn.decomposition import TruncatedSVD, LatentDirichletAllocation
+from sklearn.metrics.pairwise import cosine_similarity
+
+from covidfaq.evaluating.model.model_evaluation_interface import (
+    ModelEvaluationInterface,
+)
+
+def build_isolator(chars):
```
Comment (mirkobronzi) [inline covidfaq/evaluating/model/lda.py:10]: duplicate in __init__.py - we can use that one.

File: covidfaq/evaluating/model/lsa.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import numpy as np
+from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
+from sklearn.decomposition import TruncatedSVD
+from sklearn.metrics.pairwise import cosine_similarity
+
+from covidfaq.evaluating.model.model_evaluation_interface import (
+    ModelEvaluationInterface,
+)
+
+def build_isolator(chars):
```
Comment (mirkobronzi) [inline covidfaq/evaluating/model/lsa.py:10]: duplicate in init.py - we can use that one.

File: covidfaq/evaluating/model/tfidf.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import numpy as np
+from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
+from sklearn.decomposition import TruncatedSVD
+from sklearn.metrics.pairwise import cosine_similarity
+
+from covidfaq.evaluating.model.model_evaluation_interface import (
+    ModelEvaluationInterface,
+)
+
+def build_isolator(chars):
```
Comment (mirkobronzi) [inline covidfaq/evaluating/model/tfidf.py:10]: duplicate in init.py - we can use that one.

File: covidfaq/evaluating/top_k_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
```
Comment (mirkobronzi) [inline covidfaq/evaluating/top_k_evaluator.py:1]: we can just add an option in evaluator.py to peform top-k as well. so we can avoid duplicating the script - and we can run all in one shot..

File: covidfaq/evaluating/model/es_enhanced_reranker.py (hunk 1 above)
Comment (xhlulu) [reply covidfaq/evaluating/model/es_enhanced_reranker.py:1]: Good idea.

File: covidfaq/evaluating/top_k_evaluator.py (hunk 5 above)
Comment (xhlulu) [reply covidfaq/evaluating/top_k_evaluator.py:1]: Done

**Comment by xhlulu** [comment]
@mirkobronzi thank you for the thorough review! I brought all the changes you requested. Additionally, i also added a readme for clarity.
